doc/multicore.md

```diff
@@ -0,0 +1,521 @@
+
+```
+x     : [AB]5----------------->
+ready : [AB]false------------->
+```
+
+This means that e.g. `ready`'s timeline contains just the initial value `false`,
+and both processors are at that point on the timeline.
+The set of positions over all timelines is called the CPU's *frontier*.
+For example, the column of `A`s above is `A`'s frontier.
+
+When a processor reads from a memory location, it may get back any value at or after its current position.
+At the moment, there is only one value on each timeline, so reads will just return that.
+
+When a processor writes to a memory location, it adds the new value somewhere to the right of its current position,
+and then moves to that position.
+So once the first branch has performed the `x := !x * 2` step, we have:
+
+```
+x     : [B]5--------[A]10----->
+ready : [AB]false------------->
+```
+
+If `A` reads from `x` at this point it will certainly read back the `10` it just wrote.
+But if `B` were to read from `x`, it might see `5` or it might see `10`.
+Concretely, this corresponds to possibilities such as `10` being in the computer's main RAM
+but `5` being in `B`'s cache.
+
+The first branch then sets `ready := true`:
+
+```
+x     : [B]5--------[A]10----->
+ready : [B]false---[A]true---->
+```
+
+I've shown the `true` slightly left of the `10` just to emphasise that the timelines are independent;
+a write can go anywhere to the right of the current location.
+
+If the `while` loop keeps reading `false` (which it can) then the program will not terminate.
+If it does read `true`, it will then read and display `x`, which it might see as either `5` or `10`.
+This explains the three possible behaviours of the program on a real computer,
+without having to think about caches, out-of-order execution, or compiler optimisations.
+
+### Atomic Locations
+
+OCaml also provides special "atomic" locations.
+An atomic location just stores a single value; it does not have a timeline.
+We could fix our program by replacing the `ref` cells with atomics, like this:
+
+```ocaml
+let test run_in_new_domain =
+  let x = Atomic.make 5 in
+  let ready = Atomic.make false in
+  Fibre.both
+    (fun () -> run_in_new_domain (fun () ->
+       Atomic.set x (Atomic.get x * 2);
+       Atomic.set ready true
+    ))
+    (fun () -> run_in_new_domain (fun () ->
+       while not (Atomic.get ready) do () done;
+       traceln "x = %d" (Atomic.get x)
+    ))
+```
+
+This version will always produce the expected result:
+
+```ocaml
+# run test;;
++x = 10
+- : unit = ()
+```
+
+However, using atomics everywhere is slow.
+For example, the first `Atomic.get x` might require loading `x` from RAM, even if it was already in the cache.
+
+We can solve this by relying on a useful feature of atomics: every atomic also has a frontier of its own (a location on every non-atomic location's timeline).
+Writing to an atomic updates its frontier with information from the writing CPU's frontier
+(so it's at least as up-to-date as the writer).
```

Comment:
I initially read this as updating only the frontier of the atomic.
Looking at the PLDI'18-paper https://kcsrk.info/papers/pldi18-memory.pdf
I understand rule (While-AT) in Fig.1c as updating both frontiers.
This is also how I understand KC's example of an atomic write on sl.21-22
  https://speakerdeck.com/kayceesrk/bounding-data-races-in-space-and-time?slide=78
which updates both the red frontier for the atomic `A` and thread 2's frontier for variable `b`.

I'm fairly new to the memory-model though, so take this with a grain of salt.
If my understanding is correct, you could consider rephrasing to something like:
>  "Writing to an atomic updates both the atomic's frontier and the writing CPU's frontier
>   to contain the most up-to-date entries for each."

For the example it shouldn't change anything though.

Reply:
I think you're right. I've fixed the text. I wonder why it's like that, though?